cmd/zdb/zdb.c

```diff
@@ -4191,6 +4201,170 @@ same_metaslab(spa_t *spa, uint64_t vdev, uint64_t off1, uint64_t off2)
+ */
+typedef struct one_histo {
+	char *name;
+	uint64_t *count;
+	uint64_t *len;
+	uint64_t cumulative;
+} one_histo_t;
+
+/*
+ * The number of separate histograms processed for psize, lsize and asize.
+ */
+#define	NUM_HISTO 3
+
+/*
+ * This routine will create a fixed column size output of three different
+ * histograms showing by blocksize of 512 - 2^ SPA_MAX_FOR_16M
+ * the count, length and cumulative length of the psize, lsize and
+ * asize blocks.
+ *
+ * All three types of blocks are listed on a single line
+ *
+ * By default the table is printed in nicenumber format (e.g. 123K) but
+ * if the '-P' parameter is specified then the full raw number (parseable)
+ * is printed out.
+ */
+static void
+dump_size_histograms(zdb_cb_t *zcb)
+{
+	/*
+	 * A temporary buffer that allows us to convert a number into
+	 * a string using zdb_nicenumber to allow either raw or human
+	 * readable numbers to be output.
+	 */
+	char numbuf[32];
+
+	/*
+	 * Define titles which are used in the headers of the tables
+	 * printed by this routine.
+	 */
+	const char blocksize_title1[] = "block";
+	const char blocksize_title2[] = "size";
+	const char count_title[] = "Count";
+	const char length_title[] = "Length";
+	const char cumulative_title[] = "Cum.";
+
+	/*
+	 * Setup the histogram arrays (psize, lsize, and asize).
+	 */
+	one_histo_t parm_histo[NUM_HISTO];
+
+	parm_histo[0].name = "psize";
+	parm_histo[0].count = zcb->zcb_psize_count;
+	parm_histo[0].len = zcb->zcb_psize_len;
+	parm_histo[0].cumulative = 0;
+
+	parm_histo[1].name = "lsize";
+	parm_histo[1].count = zcb->zcb_lsize_count;
+	parm_histo[1].len = zcb->zcb_lsize_len;
+	parm_histo[1].cumulative = 0;
+
+	parm_histo[2].name = "asize";
+	parm_histo[2].count = zcb->zcb_asize_count;
+	parm_histo[2].len = zcb->zcb_asize_len;
+	parm_histo[2].cumulative = 0;
+
+	/*
+	 * Print the first line titles
+	 */
+	if (dump_opt['P'])
+		(void) printf("\n%s\t", blocksize_title1);
+	else
+		(void) printf("\n%7s   ", blocksize_title1);
+
+	for (int j = 0; j < NUM_HISTO; j++) {
+		if (dump_opt['P']) {
+			if (j < NUM_HISTO - 1) {
+				(void) printf("%s\t\t\t", parm_histo[j].name);
+			} else {
+				/* Don't print trailing spaces */
+				(void) printf("  %s", parm_histo[j].name);
+			}
+		} else {
+			if (j < NUM_HISTO - 1) {
+				/* Left aligned strings in the output */
+				(void) printf("%-7s              ",
+				    parm_histo[j].name);
+			} else {
+				/* Don't print trailing spaces */
+				(void) printf("%s", parm_histo[j].name);
+			}
+		}
+	}
+	(void) printf("\n");
+
+	/*
+	 * Print the second line titles
+	 */
+	if (dump_opt['P']) {
+		(void) printf("%s\t", blocksize_title2);
+	} else {
+		(void) printf("%7s ", blocksize_title2);
+	}
+
+	for (int i = 0; i < NUM_HISTO; i++) {
+		if (dump_opt['P']) {
+			(void) printf("%s\t%s\t%s\t",
+			    count_title, length_title, cumulative_title);
+		} else {
+			(void) printf("%7s%7s%7s",
+			    count_title, length_title, cumulative_title);
+		}
+	}
+	(void) printf("\n");
```

Comment:
Adding a section heading here would be good to explain what this table is showing us.  How about "Block Size Histogram" with a blank line above and below.

```
Block Size Histogram

  block   psize                lsize                asize
   size   Count Length   Cum.  Count Length   Cum.  Count Length   Cum.
    512:  25.3K  12.7M  12.7M  25.3K  12.7M  12.7M      0      0      0
     1K:  38.0K  46.0M  58.7M  38.0K  46.0M  58.7M      0      0      0
     2K:  39.5K   104M   163M  39.5K   104M   163M      0      0      0
     4K:  45.3K   236M   399M  36.7K   202M   365M   101K   405M   405M
     8K:  29.8K   331M   730M  29.9K   332M   697M  66.9K   605M  1010M
    16K:  19.9K   440M  1.14G  26.6K   548M  1.22G  27.1K   551M  1.52G
    32K:  10.7K   470M  1.60G  10.5K   462M  1.67G  12.5K   523M  2.04G
    64K:  4.14K   356M  1.95G  4.14K   356M  2.01G  4.81K   404M  2.43G
   128K:   175K  21.9G  23.9G   177K  22.2G  24.2G   175K  21.9G  24.3G
   256K:      0      0  23.9G      0      0  24.2G      0      0  24.3G
   512K:      0      0  23.9G      0      0  24.2G      0      0  24.3G
     1M:      0      0  23.9G      0      0  24.2G      0      0  24.3G
     2M:      0      0  23.9G      0      0  24.2G      0      0  24.3G
     4M:      0      0  23.9G      0      0  24.2G      0      0  24.3G
     8M:      0      0  23.9G      0      0  24.2G      0      0  24.3G
    16M:      0      0  23.9G      0      0  24.2G      0      0  24.3G
```

Reply:
`	parm_histo[2].cumulative = 0;


	(void) printf("\nBlock Size Histogram\n\n");
	/*
	 * Print the first line titles
	 */
Done!`